fix: wrap returned errors with context (#10592)

AvhiMaz opened the pull request:
### Summary

This PR wraps returned errors with context in several helper functions, improving error messages for debugging.

### Motivation

Addresses issue #10592 — plain `return err` statements make it harder to understand what failed.

### Checklist

* [x] This is a bug fix
* [x] The title of the PR states what changed and the related issue number (#10592)
* [x] Signed off all commits as required by DCO
* [x] Unit tests pass (no tests broken)
* [x] Build is green
* [ ] Documentation updates not required

reggie-k commented on applicationset/controllers/applicationset_controller.go:
```diff
@@ -1422,8 +1423,9 @@ func (r *ApplicationSetReconciler) migrateStatus(ctx context.Context, appset *ar
 			err := r.Client.Status().Update(ctx, updatedAppset)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed to update AppSet status: %w", err)
```
```suggestion
				return fmt.Errorf("failed to update AppSet status with ApplicationStatus during migrateStatus: %w", err)
```

reggie-k commented on applicationset/controllers/applicationset_controller.go:
```diff
@@ -1469,7 +1471,7 @@ func (r *ApplicationSetReconciler) updateResourcesStatus(ctx context.Context, lo
 		err := r.Client.Status().Update(ctx, updatedAppset)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to update AppSet status: %w", err)
```
```suggestion
			return fmt.Errorf("failed to update AppSet status with new status resources: %w", err)
```

reggie-k commented on util/localconfig/file_perm_windows.go:
```diff
@@ -8,7 +8,7 @@ import (
 
 func getFilePermission(fi os.FileInfo) error {
-	if fi.Mode().Perm() == 0666 || fi.Mode().Perm() == 0444 {
+	if fi.Mode().Perm() == 0o666 || fi.Mode().Perm() == 0o444 {
```
I think we should focus only on the error handling in this PR, let's add this change separately.

reggie-k reviewed (changes requested):
Thanks for the PR!
One issue I found with the proposed changes is that the some of the newly wrapped K8s error are not sanitized as expected, leaking secret values as a result.
I proposed an explicit error cleanup to fix this. 
Also suggested a couple of small changes to improve differentiating similar errors in ApplicationSet controller.

ppapapetrou76 commented on controller/cache/cache.go:
```diff
@@ -670,25 +670,29 @@ func (c *liveStateCache) IsNamespaced(server *appv1.Cluster, gk schema.GroupKind
 	clusterInfo, err := c.getSyncedCluster(server)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get synced cluster %s: %w", server.Name, err)
```
`getSyncedCluster` already has a similar message when it returns the error 

<img width="795" height="212" alt="Image" src="https://github.com/user-attachments/assets/23e8172f-67c8-4d6f-8ae3-778127c450a2" />

so I'm not sure if we want to add this twice

ppapapetrou76 reviewed (commented):
I don't fully agree with all these error wrapping because the added message doesn't provided any additional value 

The undrying error already has the message explaining what's wrong so my recommendation is to add wrapping and extra context only when it's missing

AvhiMaz commented:
hi @reggie-k, all test cases have passed. can you please merge it now?

reggie-k commented on gitops-engine/agent/main.go:
```diff
@@ -70,7 +70,7 @@ func (s *settings) parseManifests(ctx context.Context) ([]*unstructured.Unstruct
 		if err := filepath.Walk(filepath.Join(s.repoPath, s.paths[i]), func(path string, info os.FileInfo, err error) error {
 			if err != nil {
-				return err
+				return fmt.Errorf("error walking path %s: %w", path, err)
```
I'm not sure about the security impact of printing the path here. I think it is better to remove it.

AvhiMaz committed:
fix: address PR review feedback on error wrapping

- Include destination cluster information in namespace permission error message in sync.go for better debugging context
- Remove sensitive path information from error message in agent/main.go to prevent information leakage

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

ppapapetrou76 reviewed (commented):
I want to leave one more general comment regarding wrapping errors

The rule of thumb is to return errors without wrapping from unexported functions and return errors with context wrapping from exported functions. 

Also it would be great to start applying this rule per component.

This PR touches a lot of components - gitpops-engine might be a great point to limit this PR

nitishfy reviewed (commented):
No activity, feel free to reopen
